crypto/s2n_certificate.c

```diff
@@ -160,36 +159,51 @@ struct s2n_cert_chain_and_key *s2n_cert_chain_and_key_new(void)
     chain_and_key = (struct s2n_cert_chain_and_key *)(void *)chain_and_key_mem.data;
 
     /* Allocate the memory for the chain and key */
-    GUARD_PTR(s2n_alloc(&cert_chain_mem, sizeof(struct s2n_cert_chain)));
+    if (s2n_alloc(&cert_chain_mem, sizeof(struct s2n_cert_chain))) {
+        goto cleanup_last;
+    }
     chain_and_key->cert_chain = (struct s2n_cert_chain *)(void *)cert_chain_mem.data;
 
-    GUARD_PTR(s2n_alloc(&pkey_mem, sizeof(s2n_cert_private_key)));
+    if (s2n_alloc(&pkey_mem, sizeof(s2n_cert_private_key))) {
+        goto cleanup_second;
+    }
     chain_and_key->private_key = (s2n_cert_private_key *)(void *)pkey_mem.data;
 
     chain_and_key->cert_chain->head = NULL;
-    GUARD_PTR(s2n_pkey_zero_init(chain_and_key->private_key));
+    if (s2n_pkey_zero_init(chain_and_key->private_key)) {
+        goto cleanup_all;
+    }
     memset(&chain_and_key->ocsp_status, 0, sizeof(chain_and_key->ocsp_status));
     memset(&chain_and_key->sct_list, 0, sizeof(chain_and_key->sct_list));
     chain_and_key->cn_names = s2n_array_new(sizeof(struct s2n_blob));
     if (!chain_and_key->cn_names) {
-        return NULL;
+        goto cleanup_all;
     }
 
     chain_and_key->san_names = s2n_array_new(sizeof(struct s2n_blob));
     if (!chain_and_key->san_names) {
-        return NULL;
+        goto cleanup_all;
     }
 
     chain_and_key->context = NULL;
 
     return chain_and_key;
+    cleanup_all:
```

Comment:
Do we need multiple labels since `s2n_free` should check that the pointer is nonnull?

Reply:
Looking at the code, it looks like we'll get the `s2n_errno` set if we do this.

https://github.com/awslabs/s2n/blob/981ef47b25251c03abf5d4d8a0284e6c3cd5c0d9/utils/s2n_mem.c#L277

I wonder if it would be worth adding something like this to `s2n_free`:

```
if (b == NULL) {
    return S2N_SUCCESS;
}
```

Reply:
That seems reasonable. I wonder if there's a reason why we don't already do that, or it was just a miss.